File: src/render.ts
/** Deterministic byte renderers (spec §9.3, §10). Pure serializers over closed objects. */

import { blobHash, canonicalHash, canonicalize } from "./canon.js";
import { writePack } from "./zipio.js";
import type { Signer } from "./sign.js";

export const NOTICE1 = "Issuer-attested test observations; not accredited standards compliance";
export const NOTICE2 = "Current status must be checked; this PDF is a dated record";
export const RETENTION = "Retention: private evidence 180 days; public record 2555 days minimum.";

export function utc(t: number): string {
  return new Date(t * 1000).toISOString().replace(/\.\d{3}Z$/, "Z");
}

export function fields(cert: any, result: any, release: any): string[] {
  const c = cert.body, rel = release.body;
  const lines = [
    "EvalSeal assessment",
    "Qualification: " + c.qualification,
    "Band: " + c.band,
    "Product: " + c.product_label + " / " + c.product_version,
    "Issuer key: " + cert.key_id,
    "Issued: " + utc(c.issued_at),
    "Expiry: " + utc(c.expires_at),
    "Target: " + c.target_hash,
    "Release: " + c.release_hash,
  ];
  lines.push(...rel.suites.map((s: any) => "Suite: " + s.family + " / " + s.version));
  lines.push("Runner: " + rel.runner_image_hash, "Oracle: " + rel.oracle_hash);
  lines.push(...result.families.map((f: any) =>
    `${f.family}: adversarial ${f.adversarial_pass}/30; benign ${f.benign_pass}/10; critical failures ${f.critical_fail}/5`));
  return [...lines,
    "Reproducibility: " + c.reproducibility_hash,
    "Evidence head: " + c.evidence_head,
    ...c.scope_lines,
    RETENTION,
    "Verification path: /certificates/" + c.certificate_id,
  ];
}

const safe = (s: string) =>
  [...s].map(ch => {
    const o = ch.codePointAt(0)!;
    return o >= 32 && o <= 126 ? ch : `[U+${o.toString(16).toUpperCase().padStart(4, "0")}]`;
  }).join("");

/** textwrap.wrap(line, width, break_long_words=True, break_on_hyphens=False) */
export function textWrap(line: string, width: number): string[] {
  const chunks = line.split(/(\s+)/).filter(c => c.length > 0).reverse();
  const lines: string[] = [];
  while (chunks.length) {
    const cur: string[] = [];
    let curLen = 0;
    if (chunks[chunks.length - 1]!.trim() === "" && lines.length) chunks.pop();
    while (chunks.length) {
      const l = chunks[chunks.length - 1]!.length;
      if (curLen + l <= width) { cur.push(chunks.pop()!); curLen += l; }
      else break;
    }
    if (chunks.length && chunks[chunks.length - 1]!.length > width) {
      const spaceLeft = Math.max(1, width - curLen);
      const chunk = chunks[chunks.length - 1]!;
      cur.push(chunk.slice(0, spaceLeft));
      chunks[chunks.length - 1] = chunk.slice(spaceLeft);
    }
    if (cur.length && cur[cur.length - 1]!.trim() === "") cur.pop();
    if (cur.length) lines.push(cur.join(""));
  }
  return lines;
}

export function renderPdf(cert: any, result: any, release: any): Buffer {
  const lines: string[] = [];
  for (const line of fields(cert, result, release)) {
    lines.push(...textWrap(safe(line), 80));
  }
  const chunks: string[][] = [];
  for (let i = 0; i < lines.length; i += 55) chunks.push(lines.slice(i, i + 55));
  const objects = new Map<number, Buffer>();
  objects.set(1, Buffer.from("<< /Type /Catalog /Pages 2 0 R >>"));
  objects.set(3, Buffer.from("<< /Type /Font /Subtype /Type1 /BaseFont /Courier >>"));
  const kids = chunks.map((_, i) => `${4 + 2 * i} 0 R`).join(" ");
  objects.set(2, Buffer.from(`<< /Type /Pages /Kids [${kids}] /Count ${chunks.length} >>`));
  const esc = (s: string) => s.replace(/\\/g, "\\\\").replace(/\(/g, "\\(").replace(/\)/g, "\\)");
  chunks.forEach((chunk, i) => {
    const page = 4 + 2 * i, content = 5 + 2 * i;
    const text = [...chunk, `Page ${i + 1}/${chunks.length}`, NOTICE1, NOTICE2];
    const stream = Buffer.from(
      "BT /F1 10 Tf 12 TL 40 802 Td\n" + text.map(s => "(" + esc(s) + ") Tj").join("\nT*\n") + "\nET\n", "ascii");
    objects.set(page, Buffer.from(
      `<< /Type /Page /Parent 2 0 R /MediaBox [0 0 595 842] /Resources << /Font << /F1 3 0 R >> >> /Contents ${content} 0 R >>`));
    objects.set(content, Buffer.concat([Buffer.from(`<< /Length ${stream.length} >>\nstream\n`), stream, Buffer.from("endstream")]));
  });
  const parts: Buffer[] = [Buffer.from("%PDF-1.4\n")];
  const offsets: number[] = [0];
  let pos = 9;
  for (let n = 1; n <= objects.size; n++) {
    offsets.push(pos);
    const obj = Buffer.concat([Buffer.from(`${n} 0 obj\n`), objects.get(n)!, Buffer.from("\nendobj\n")]);
    parts.push(obj);
    pos += obj.length;
  }
  const xref = pos;
  const tail = `xref\n0 ${offsets.length}\n0000000000 65535 f \n` +
    offsets.slice(1).map(o => `${String(o).padStart(10, "0")} 00000 n \n`).join("") +
    `trailer\n<< /Size ${offsets.length} /Root 1 0 R >>\nstartxref\n${xref}\n%%EOF\n`;
  parts.push(Buffer.from(tail));
  return Buffer.concat(parts);
}

export function packFiles(cert: any, result: any, release: any, target: any, audit: any[],
                          checkpoint: any, keyring: any, cases: any[],
                          disclosure = "public-redacted", evidence?: any, transcript?: Buffer) {
  const files = new Map<string, Buffer>([
    ["certificate.json", canonicalize(cert)],
    ["release.json", canonicalize(release)],
    ["cases.json", canonicalize(cases)],
    ["target.json", canonicalize(target)],
    ["result.json", canonicalize(result)],
    ["audit.jsonl", Buffer.concat(audit.map(e => Buffer.concat([canonicalize(e), Buffer.from("\n")])))],
    ["checkpoint.json", canonicalize(checkpoint)],
    ["keyring.json", canonicalize(keyring)],
    ["scope.txt", Buffer.from([...cert.body.scope_lines, RETENTION].join("\n") + "\n")],
    ["report.pdf", renderPdf(cert, result, release)],
  ]);
  if (disclosure === "private-full") {
    if (evidence === undefined || transcript === undefined) {
      throw new Error("private-full pack requires evidence and transcript");
    }
    files.set("evidence.json", canonicalize(evidence));
    files.set("transcript.bin", transcript);
  }
  const media = (n: string) =>
    n.endsWith(".pdf") ? "application/pdf"
      : n.endsWith(".txt") || n.endsWith(".jsonl") ? "text/plain"
      : n.endsWith(".bin") ? "application/octet-stream"
      : "application/json";
  const manifest = {
    schema: "evalseal-pack/1",
    certificate_hash: canonicalHash(cert as never),
    files: [...files.entries()].sort(([a], [b]) => a < b ? -1 : 1).map(([n, b]) =>
      ({ path: n, hash: blobHash(b), bytes: b.length, media_type: media(n) })),
    disclosure,
  };
  return { files, manifest };
}

export function packZip(cert: any, result: any, release: any, target: any, audit: any[],
                        checkpoint: any, keyring: any, cases: any[], signer?: Signer,
                        disclosure = "public-redacted", evidence?: any, transcript?: Buffer): Buffer {
  const { files, manifest } = packFiles(cert, result, release, target, audit, checkpoint,
    keyring, cases, disclosure, evidence, transcript);
  files.set("manifest.json", canonicalize(signer ? signer("pack", manifest) : manifest));
  const sorted = new Map([...files.entries()].sort(([a], [b]) => (a < b ? -1 : 1)));
  return writePack(sorted);
}

const htmlEscape = (s: string) => s.replace(/&/g, "&amp;").replace(/</g, "&lt;")
  .replace(/>/g, "&gt;").replace(/"/g, "&quot;").replace(/'/g, "&#x27;");

export function renderPage(cert: any, result: any, status: any, release: any): Buffer {
  const text = [...fields(cert, result, release),
    "Status: " + status.body.state,
    "Status at: " + utc(status.body.as_of),
    NOTICE1, NOTICE2];
  return Buffer.from(
    '<!doctype html><html lang="en"><meta charset="utf-8"><title>EvalSeal assessment</title>' +
    "<main><h1>EvalSeal assessment</h1>" +
    text.map(s => "<p>" + htmlEscape(s) + "</p>").join("") +
    "</main></html>");
}

export function badgeLines(cert: any, status: any): string[] {
  const c = cert.body, s = status.body;
  const primary = c.band === "F" ? "NOT CERTIFIED — F"
    : s.state !== "ACTIVE" ? s.state
    : c.band + " CERTIFIED";
  return [
    "EvalSeal | " + primary,
    c.product_label + " / " + c.product_version,
    "Target: " + c.target_hash,
    "Issued: " + utc(c.issued_at),
    "Expiry: " + utc(c.expires_at),
    "Status at: " + utc(s.as_of),
    ...c.scope_lines,
    "/certificates/" + c.certificate_id,
  ];
}

function badgeSvg(lines: string[]): Buffer {
  const wrapped = lines.flatMap(t => textWrap(t, 100));
  const nodes = wrapped.map((t, i) =>
    `<text x="20" y="${30 + 18 * i}" font-family="monospace" font-size="12">${htmlEscape(t)}</text>`).join("");
  return Buffer.from(
    `<svg xmlns="http://www.w3.org/2000/svg" width="960" height="${50 + 18 * wrapped.length}" role="img">` +
    `<title>EvalSeal scoped assessment</title>${nodes}</svg>`);
}

export function renderBadge(cert: any, status: any): Buffer {
  return badgeSvg(badgeLines(cert, status));
}

export function renderBadgeUnavailable(): Buffer {
  return Buffer.from(
    '<svg xmlns="http://www.w3.org/2000/svg" width="960" height="68" role="img">' +
    "<title>EvalSeal scoped assessment</title>" +
    '<text x="20" y="30" font-family="monospace" font-size="12">Record unavailable</text>' +
    "</svg>");
}

export function renderBadgeUnknown(cert: any, statusTime: number): Buffer {
  const c = cert.body;
  return badgeSvg([
    "EvalSeal | STATUS UNKNOWN",
    c.product_label + " / " + c.product_version,
    "Target: " + c.target_hash,
    "Issued: " + utc(c.issued_at),
    "Expiry: " + utc(c.expires_at),
    "Status at: " + utc(statusTime),
    ...c.scope_lines,
    "/certificates/" + c.certificate_id,
  ]);
}

export function renderLeaderboard(page: any, release: any): Buffer {
  const rows = page.items.map((r: any) =>
    "<tr>" + ["label", "state", "band", "status"].map(k => `<td>${htmlEscape(String(r[k]))}</td>`).join("") + "</tr>").join("");
  const scope = release.body.scope_lines.map((s: string) => `<li>${htmlEscape(s)}</li>`).join("");
  return Buffer.from(
    '<!doctype html><html lang="en"><meta charset="utf-8"><title>EvalSeal leaderboard</title>' +
    '<main><h1>EvalSeal leaderboard</h1><ul>' + scope + "</ul>" +
    '<table><caption>Latest public attempts</caption><thead>' +
    "<tr><th>Product</th><th>State</th><th>Band</th><th>Status</th></tr></thead>" +
    "<tbody>" + rows + "</tbody></table></main></html>");
}

File: python/evalseal/render.py
"""Deterministic byte renderers (spec §9.3, §10).

These are pure serializers over closed objects — the reference contract that
fixes PDF field names, archive metadata, and minimal HTML/SVG output.
"""

from __future__ import annotations

import datetime
import html
import io
import textwrap
import zipfile

from .canon import blob_hash, canonical_hash, canonicalize
from .sign import sign_envelope

NOTICE1 = "Issuer-attested test observations; not accredited standards compliance"
NOTICE2 = "Current status must be checked; this PDF is a dated record"
RETENTION = "Retention: private evidence 180 days; public record 2555 days minimum."


def utc(t: int) -> str:
    return datetime.datetime.fromtimestamp(t, datetime.timezone.utc).isoformat().replace("+00:00", "Z")


def fields(cert: dict, result: dict, release: dict) -> list:
    c, rel = cert["body"], release["body"]
    lines = [
        "EvalSeal assessment",
        "Qualification: " + c["qualification"],
        "Band: " + c["band"],
        "Product: " + c["product_label"] + " / " + c["product_version"],
        "Issuer key: " + cert["key_id"],
        "Issued: " + utc(c["issued_at"]),
        "Expiry: " + utc(c["expires_at"]),
        "Target: " + c["target_hash"],
        "Release: " + c["release_hash"],
    ]
    lines += ["Suite: " + s["family"] + " / " + s["version"] for s in rel["suites"]]
    lines += ["Runner: " + rel["runner_image_hash"], "Oracle: " + rel["oracle_hash"]]
    lines += [
        f'{f["family"]}: adversarial {f["adversarial_pass"]}/30; benign {f["benign_pass"]}/10; critical failures {f["critical_fail"]}/5'
        for f in result["families"]
    ]
    return lines + [
        "Reproducibility: " + c["reproducibility_hash"],
        "Evidence head: " + c["evidence_head"],
    ] + c["scope_lines"] + [
        RETENTION,
        "Verification path: /certificates/" + c["certificate_id"],
    ]


def _safe(s: str) -> str:
    return "".join(ch if 32 <= ord(ch) <= 126 else f"[U+{ord(ch):04X}]" for ch in s)


def render_pdf(cert: dict, result: dict, release: dict) -> bytes:
    """Uncompressed PDF 1.4 / Courier 10pt / A4 / 40pt margins (spec §9.3)."""
    lines = [
        part
        for line in fields(cert, result, release)
        for part in textwrap.wrap(_safe(line), 80, break_long_words=True, break_on_hyphens=False)
    ]
    chunks = [lines[i : i + 55] for i in range(0, len(lines), 55)]
    objects = {1: b"<< /Type /Catalog /Pages 2 0 R >>", 3: b"<< /Type /Font /Subtype /Type1 /BaseFont /Courier >>"}
    kids = " ".join(f"{4 + 2 * i} 0 R" for i in range(len(chunks)))
    objects[2] = f"<< /Type /Pages /Kids [{kids}] /Count {len(chunks)} >>".encode()
    escape = lambda s: s.replace("\\", "\\\\").replace("(", "\\(").replace(")", "\\)")
    for i, chunk in enumerate(chunks):
        page, content = 4 + 2 * i, 5 + 2 * i
        text = chunk + [f"Page {i + 1}/{len(chunks)}", NOTICE1, NOTICE2]
        stream = ("BT /F1 10 Tf 12 TL 40 802 Td\n" + "\nT*\n".join("(" + escape(s) + ") Tj" for s in text) + "\nET\n").encode("ascii")
        objects[page] = f"<< /Type /Page /Parent 2 0 R /MediaBox [0 0 595 842] /Resources << /Font << /F1 3 0 R >> >> /Contents {content} 0 R >>".encode()
        objects[content] = f"<< /Length {len(stream)} >>\nstream\n".encode() + stream + b"endstream"
    out, offsets = bytearray(b"%PDF-1.4\n"), [0]
    for number in range(1, len(objects) + 1):
        offsets.append(len(out))
        out.extend(f"{number} 0 obj\n".encode() + objects[number] + b"\nendobj\n")
    xref = len(out)
    out.extend(f"xref\n0 {len(offsets)}\n0000000000 65535 f \n".encode())
    out.extend("".join(f"{offset:010d} 00000 n \n" for offset in offsets[1:]).encode())
    out.extend(f"trailer\n<< /Size {len(offsets)} /Root 1 0 R >>\nstartxref\n{xref}\n%%EOF\n".encode())
    return bytes(out)


def pack_files(cert, result, release, target, audit, checkpoint, keyring, cases,
               disclosure="public-redacted", evidence=None, transcript=None) -> dict:
    """The exact attestation-pack file set (spec §9.2)."""
    files = {
        "certificate.json": canonicalize(cert),
        "release.json": canonicalize(release),
        "cases.json": canonicalize(cases),
        "target.json": canonicalize(target),
        "result.json": canonicalize(result),
        "audit.jsonl": b"".join(canonicalize(e) + b"\n" for e in audit),
        "checkpoint.json": canonicalize(checkpoint),
        "keyring.json": canonicalize(keyring),
        "scope.txt": ("\n".join(cert["body"]["scope_lines"]) + "\n" + RETENTION + "\n").encode(),
        "report.pdf": render_pdf(cert, result, release),
    }
    if disclosure == "private-full":
        if evidence is None or transcript is None:
            raise ValueError("private-full pack requires evidence and transcript")
        files["evidence.json"] = canonicalize(evidence)
        files["transcript.bin"] = transcript
    media = lambda name: (
        "application/pdf" if name.endswith(".pdf")
        else "text/plain" if name.endswith((".txt", ".jsonl"))
        else "application/octet-stream" if name.endswith(".bin")
        else "application/json"
    )
    manifest = {
        "schema": "evalseal-pack/1",
        "certificate_hash": canonical_hash(cert),
        "files": [
            {"path": n, "hash": blob_hash(b), "bytes": len(b), "media_type": media(n)}
            for n, b in sorted(files.items())
        ],
        "disclosure": disclosure,
    }
    return files, manifest


def pack_zip(cert, result, release, target, audit, checkpoint, keyring, cases,
             signer=None, disclosure="public-redacted",
             evidence=None, transcript=None) -> bytes:
    """STORE archive, sorted ASCII paths, DOS epoch, unix 0644 (spec §9.3).

    signer is a callable (kind, body) -> Signed envelope. Unsigned manifests
    are produced only for local simulation tooling.
    """
    files, manifest = pack_files(
        cert, result, release, target, audit, checkpoint, keyring, cases,
        disclosure, evidence, transcript)
    files["manifest.json"] = canonicalize(signer("pack", manifest)) if signer else canonicalize(manifest)
    output = io.BytesIO()
    with zipfile.ZipFile(output, "w", compression=zipfile.ZIP_STORED) as archive:
        for name, data in sorted(files.items()):
            info = zipfile.ZipInfo(name, (1980, 1, 1, 0, 0, 0))
            info.create_system = 3
            info.external_attr = 0o100644 << 16
            archive.writestr(info, data)
    return output.getvalue()


def render_page(cert, result, status, release) -> bytes:
    text = fields(cert, result, release) + [
        "Status: " + status["body"]["state"],
        "Status at: " + utc(status["body"]["as_of"]),
        NOTICE1,
        NOTICE2,
    ]
    return (
        '<!doctype html><html lang="en"><meta charset="utf-8"><title>EvalSeal assessment</title>'
        "<main><h1>EvalSeal assessment</h1>"
        + "".join("<p>" + html.escape(s) + "</p>" for s in text)
        + "</main></html>"
    ).encode()


def badge_lines(cert, status) -> list:
    """Ordered visible strings, independent of SVG styling (spec §14.1)."""
    c, s = cert["body"], status["body"]
    if c["band"] == "F":
        primary = "NOT CERTIFIED — F"
    elif s["state"] != "ACTIVE":
        primary = s["state"]
    else:
        primary = c["band"] + " CERTIFIED"
    return [
        "EvalSeal | " + primary,
        c["product_label"] + " / " + c["product_version"],
        "Target: " + c["target_hash"],
        "Issued: " + utc(c["issued_at"]),
        "Expiry: " + utc(c["expires_at"]),
        "Status at: " + utc(s["as_of"]),
    ] + c["scope_lines"] + ["/certificates/" + c["certificate_id"]]


def render_badge(cert, status) -> bytes:
    text = badge_lines(cert, status)
    lines = [p for t in text for p in textwrap.wrap(t, 100, break_long_words=True, break_on_hyphens=False)]
    nodes = "".join(
        f'<text x="20" y="{30 + 18 * i}" font-family="monospace" font-size="12">{html.escape(t)}</text>'
        for i, t in enumerate(lines)
    )
    return (
        f'<svg xmlns="http://www.w3.org/2000/svg" width="960" height="{50 + 18 * len(lines)}" role="img">'
        f"<title>EvalSeal scoped assessment</title>{nodes}</svg>"
    ).encode()


def render_badge_unavailable() -> bytes:
    """Identical nonqualifying 404 badge for unknown or private certificates."""
    return (
        '<svg xmlns="http://www.w3.org/2000/svg" width="960" height="68" role="img">'
        "<title>EvalSeal scoped assessment</title>"
        '<text x="20" y="30" font-family="monospace" font-size="12">Record unavailable</text>'
        "</svg>"
    ).encode()


def render_badge_unknown(cert, status_time: int) -> bytes:
    """Same layout with primary text STATUS UNKNOWN when freshness fails."""
    c = cert["body"]
    lines = [
        "EvalSeal | STATUS UNKNOWN",
        c["product_label"] + " / " + c["product_version"],
        "Target: " + c["target_hash"],
        "Issued: " + utc(c["issued_at"]),
        "Expiry: " + utc(c["expires_at"]),
        "Status at: " + utc(status_time),
    ] + c["scope_lines"] + ["/certificates/" + c["certificate_id"]]
    wrapped = [p for t in lines for p in textwrap.wrap(t, 100, break_long_words=True, break_on_hyphens=False)]
    nodes = "".join(
        f'<text x="20" y="{30 + 18 * i}" font-family="monospace" font-size="12">{html.escape(t)}</text>'
        for i, t in enumerate(wrapped)
    )
    return (
        f'<svg xmlns="http://www.w3.org/2000/svg" width="960" height="{50 + 18 * len(wrapped)}" role="img">'
        f"<title>EvalSeal scoped assessment</title>{nodes}</svg>"
    ).encode()


def render_leaderboard(page, release) -> bytes:
    rows = "".join(
        "<tr>" + "".join("<td>" + html.escape(str(r[k])) + "</td>" for k in ("label", "state", "band", "status")) + "</tr>"
        for r in page["items"]
    )
    scope = "".join("<li>" + html.escape(s) + "</li>" for s in release["body"]["scope_lines"])
    return (
        '<!doctype html><html lang="en"><meta charset="utf-8"><title>EvalSeal leaderboard</title>'
        '<main><h1>EvalSeal leaderboard</h1><ul>' + scope + "</ul>"
        '<table><caption>Latest public attempts</caption><thead>'
        "<tr><th>Product</th><th>State</th><th>Band</th><th>Status</th></tr></thead>"
        "<tbody>" + rows + "</tbody></table></main></html>"
    ).encode()
